provider/ovh.go

```diff
@@ -0,0 +1,354 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+
+	"github.com/golang/sync/errgroup"
+	"github.com/ovh/go-ovh/ovh"
+	log "github.com/sirupsen/logrus"
+
+	"sigs.k8s.io/external-dns/endpoint"
+	"sigs.k8s.io/external-dns/plan"
+)
+
+const (
+	ovhDefaultTTL = 0
+)
+
+const (
+	ovhCreate = iota
+	ovhDelete
+)
+
+var (
+	// ErrRecordToMutateNotFound when ApplyChange has to update/delete and didn't found the record in the existing zone (Change with no record ID)
+	ErrRecordToMutateNotFound = errors.New("record to mutate not found in current zone")
+	// ErrNoDryRun No dry run support for the moment
+	ErrNoDryRun = errors.New("dry run not supported")
+)
+
+// OVHProvider is an implementation of Provider for OVH DNS.
+type OVHProvider struct {
+	client ovhClient
+
+	domainFilter DomainFilter
+	DryRun       bool
+}
+
+type ovhClient interface {
+	Post(string, interface{}, interface{}) error
+	Get(string, interface{}) error
+	Delete(string, interface{}) error
+}
+
+type ovhRecordFields struct {
+	FieldType string `json:"fieldType"`
+	SubDomain string `json:"subDomain"`
+	TTL       int64  `json:"ttl"`
+	Target    string `json:"target"`
+}
+
+type ovhRecord struct {
+	ovhRecordFields
+	ID   uint64 `json:"id"`
+	Zone string `json:"zone"`
+}
+
+type ovhChange struct {
+	ovhRecord
+	Action int
+}
+
+// NewOVHProvider initializes a new OVH DNS based Provider.
+func NewOVHProvider(ctx context.Context, domainFilter DomainFilter, endpoint string, generateConsumerKey bool, dryRun bool) (*OVHProvider, error) {
+	client, err := ovh.NewEndpointClient(endpoint)
+	if err != nil {
+		return nil, err
+	}
+	if generateConsumerKey {
+		ckReq := client.NewCkRequest()
+		ckReq.AddRules(ovh.ReadOnly, "/domain/zone")
+		ckReq.AddRecursiveRules(ovh.ReadWrite, "/domain/zone/*/record")
+		ckReq.AddRules([]string{"POST"}, "/domain/zone/*/refresh")
+		response, err := ckReq.Do()
+		if err != nil {
+			return nil, err
+		}
+		log.Infof("Generated consumer key: %s\n", response.ConsumerKey)
```

Comment:
I don't think this a good approach to log such critical information.

Reply:
I was hesitant to log this key, this only simplify the process of validating your OVH application.
We can replace it with the HTTP query to run, to have these informations.
Like the example here https://docs.ovh.com/gb/en/customer/first-steps-with-ovh-api/#creation-of-your-application-keys

Something like:

Run that
`
curl -XPOST -H "X-Ovh-Application: 7kbG7Bk7S9Nt7ZSV" -H "Content-type: application/json" https://eu.api.ovh.com/1.0/auth/credential  -d '{ "accessRules": [{"method": "GET", "path": "/*"}], "redirection":"https://www.mywebsite.com/"}' 
`
to have your keys

Reply:
It would be better if users need to take action first and pass it via flag, WDYT?

Reply:
Yep you're right, it's better this way.
I have removed the code who generate the consumer key and specify more informations about that in the OVH tutorial. Ok for you ?